c10/core/impl/COW.cpp

```diff
@@ -133,12 +225,25 @@ C10_API void materialize_cow_storage(StorageImpl& storage) {
     new_data_ptr = DataPtr(
         data.release(), data_ptr.get(), data.get_deleter(), data_ptr.device());
   } else {
-    TORCH_INTERNAL_ASSERT(
-        std::holds_alternative<cow::COWDeleterContext::NotLastReference>(
-            result));
     // We don't need to consume the result, it's just a shared lock ensuring
     // that the data will remain while we copy it.
-    new_data_ptr = storage.allocator()->clone(data_ptr.get(), storage.nbytes());
+    if (devices_match) {
+      new_data_ptr =
+          storage.allocator()->clone(data_ptr.get(), storage.nbytes());
+    } else {
+      new_data_ptr = clone_between_devices(
+          // NOTE: For MPS-to-CPU, `data_ptr.get()` gives an address in CPU
+          // space even though `src_device` is MPS. Likewise, for CPU-to-MPS,
+          // `data_ptr.get()` gives an address in MPS space even though the
+          // `src_device` is CPU. So both the src and dest pointers will always
+          // be in the address space of the dest device.
```

Comment:
```suggestion
          // `src_device` is CPU. So both the data_ptr and new_data_ptr pointers will always
          // be in the address space of the dest device.
```

Helpful comment, but it's clearer if the naming is consistent with the surrounding code.